sql/src/main/java/io/crate/execution/dml/upsert/ShardInsertRequest.java

```diff
@@ -0,0 +1,318 @@
+        UUID jobId,
+        SessionSettings sessionSettings,
+        Reference[] insertColumns,
+        boolean continueOnError,
+        boolean validateGeneratedColumns,
+        DuplicateKeyAction duplicateKeyAction) {
+        super(shardId, jobId);
+        this.sessionSettings = sessionSettings;
+        this.insertColumns = insertColumns;
+        this.modes = Mode.toEnumSet(continueOnError, validateGeneratedColumns, duplicateKeyAction);
+    }
+
+    public ShardInsertRequest(StreamInput in) throws IOException {
+        super(in);
+        int missingAssignmentsColumnsSize = in.readVInt();
+        Streamer[] insertValuesStreamer = null;
+        if (missingAssignmentsColumnsSize > 0) {
+            insertColumns = new Reference[missingAssignmentsColumnsSize];
+            for (int i = 0; i < missingAssignmentsColumnsSize; i++) {
+                insertColumns[i] = Reference.fromStream(in);
+            }
+            insertValuesStreamer = Symbols.streamerArray(List.of(insertColumns));
+        }
+        modes = EnumSets.unpackFromInt(in.readVInt(), Mode.class);
+        sessionSettings = new SessionSettings(in);
+        int numItems = in.readVInt();
+        items = new ArrayList<>(numItems);
+        for (int i = 0; i < numItems; i++) {
+            items.add(new ShardInsertRequest.Item(in, insertValuesStreamer));
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        Streamer[] insertValuesStreamer = null;
+        if (insertColumns != null) {
+            out.writeVInt(insertColumns.length);
+            for (Reference reference : insertColumns) {
+                Reference.toStream(reference, out);
+            }
+            insertValuesStreamer = Symbols.streamerArray(List.of(insertColumns));
+        } else {
+            out.writeVInt(0);
+        }
+        out.writeVInt(EnumSets.packToInt(modes));
+        sessionSettings.writeTo(out);
+        out.writeVInt(items.size());
+        for (ShardInsertRequest.Item item : items) {
+            item.writeTo(out, insertValuesStreamer);
+        }
+    }
+
+    @Nullable
+    @Override
+    public SessionSettings sessionSettings() {
+        return sessionSettings;
+    }
+
+    @Nullable
+    @Override
+    public Symbol[] returnValues() {
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public String[] updateColumns() {
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public Reference[] insertColumns() {
+        return insertColumns;
+    }
+
+    @Override
+    public boolean continueOnError() {
+        return modes.contains(Mode.CONTINUE_ON_ERROR);
+    }
+
+    @Override
+    public boolean validateConstraints() {
+        return modes.contains(Mode.VALIDATE_CONSTRAINTS);
+    }
+
+    @Override
+    public DuplicateKeyAction duplicateKeyAction() {
+        return Mode.getDuplicateAction(modes);
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
I personally don't see the point to add equals/hashcode to request classes, requests should be unique by memory address. Elasticsearch's request classes also don't have equals/hashcode either but since `ShardRequest` has it, the subclasses must have it.

Reply:
It can help writing streaming tests - as it allows to assert that the streamed request is equal to the original request. But otherwise I agree, equality of the requests shouldn't matter.